File: src/devtwin/adapters/go.py
"""Go ecosystem adapter."""

from __future__ import annotations

import re
from pathlib import Path

from devtwin.adapters.base import EcosystemAdapter, extract_version
from devtwin.core.models import DependencyInfo, HealthIssue, Presence, RuntimeInfo, Severity
from devtwin.core.runner import run_command, which

GO_MOD_VERSION_RE = re.compile(r"^go\s+(\d+(?:\.\d+){1,2})", re.MULTILINE)
GO_MOD_MODULE_RE = re.compile(r"^module\s+(\S+)", re.MULTILINE)


class GoAdapter(EcosystemAdapter):
    ecosystem = "go"

    def detect(self, root: Path) -> bool:
        return any((root / n).exists() for n in ("go.mod", "go.sum", "go.work"))

    def _required_version(self, root: Path) -> str | None:
        gomod = root / "go.mod"
        if not gomod.exists():
            return None
        try:
            text = gomod.read_text(encoding="utf-8", errors="ignore")
        except OSError:
            return None
        match = GO_MOD_VERSION_RE.search(text)
        return match.group(1) if match else None

    def inspect_runtime(self, root: Path) -> list[RuntimeInfo]:
        required = self._required_version(root)
        go_path = which("go")
        if go_path is None:
            return [
                RuntimeInfo(name="go", presence=Presence.NOT_INSTALLED, required_version=required)
            ]
        result = run_command([go_path, "version"], timeout=5)
        installed = extract_version(result.stdout) if result.available else None
        return [
            RuntimeInfo(
                name="go",
                presence=Presence.INSTALLED if installed else Presence.UNKNOWN,
                installed_version=installed,
                required_version=required,
                path=go_path,
                source="go.mod" if required else None,
            )
        ]

    def inspect_dependencies(self, root: Path) -> DependencyInfo | None:
        gomod = root / "go.mod"
        if not gomod.exists():
            return None
        try:
            text = gomod.read_text(encoding="utf-8", errors="ignore")
        except OSError:
            text = ""
        module_match = GO_MOD_MODULE_RE.search(text)
        notes = [f"module {module_match.group(1)}"] if module_match else []
        gosum = root / "go.sum"
        return DependencyInfo(
            ecosystem="go",
            manager="go modules",
            lockfile="go.sum" if gosum.exists() else None,
            lockfile_present=gosum.exists(),
            manifest_present=True,
            notes=notes,
        )

    def inspect_tests(self, root: Path) -> list[str]:
        return ["go test ./..."] if (root / "go.mod").exists() else []

    def inspect_build_commands(self, root: Path) -> list[str]:
        return ["go build ./..."] if (root / "go.mod").exists() else []

    def health_checks(self, root: Path, runtimes: list[RuntimeInfo]) -> list[HealthIssue]:
        issues = []
        for rt in runtimes:
            if rt.name == "go" and rt.presence == Presence.NOT_INSTALLED:
                issues.append(
                    HealthIssue(
                        severity=Severity.CRITICAL,
                        code="go.not_installed",
                        title="Go runtime not found",
                        message="No `go` executable was found on PATH.",
                        evidence=["which go -> not found"],
                        recommendation="Install Go from https://go.dev/dl/ or your OS package manager.",
                    )
                )
        return issues

File: mcphub/devtwin/src/adapters/go.ts
/**
 * Go ecosystem adapter.
 */

import { join } from 'node:path';

import {
  Presence,
  Severity,
  type DependencyInfo,
  type HealthIssue,
  type RuntimeInfo,
} from '../core/models.js';
import { runCommand, which } from '../core/runner.js';
import { pathExists, readTextFile } from '../system/filesystem.js';
import { EcosystemAdapter, extractVersion } from './base.js';

const GO_MOD_VERSION_RE = /^go\s+(\d+(?:\.\d+){1,2})/m;
const GO_MOD_MODULE_RE = /^module\s+(\S+)/m;

export class GoAdapter extends EcosystemAdapter {
  readonly ecosystem = 'go';

  detect(root: string): boolean {
    return ['go.mod', 'go.sum', 'go.work'].some((n) => pathExists(join(root, n)));
  }

  private requiredVersion(root: string): string | null {
    const gomod = join(root, 'go.mod');
    if (!pathExists(gomod)) {
      return null;
    }
    // readTextFile returns null instead of raising when the file is unreadable.
    const text = readTextFile(gomod);
    if (text === null) {
      return null;
    }
    const match = GO_MOD_VERSION_RE.exec(text);
    return match ? match[1]! : null;
  }

  override async inspectRuntime(root: string): Promise<RuntimeInfo[]> {
    const required = this.requiredVersion(root);
    const goPath = which('go');
    if (goPath === null) {
      return [
        {
          name: 'go',
          presence: Presence.NOT_INSTALLED,
          installed_version: null,
          required_version: required,
          path: null,
          source: null,
        },
      ];
    }
    const result = await runCommand([goPath, 'version'], { timeout: 5 });
    const installed = result.available ? extractVersion(result.stdout) : null;
    return [
      {
        name: 'go',
        presence: installed ? Presence.INSTALLED : Presence.UNKNOWN,
        installed_version: installed,
        required_version: required,
        path: goPath,
        source: required ? 'go.mod' : null,
      },
    ];
  }

  override async inspectDependencies(root: string): Promise<DependencyInfo | null> {
    const gomod = join(root, 'go.mod');
    if (!pathExists(gomod)) {
      return null;
    }
    const text = readTextFile(gomod) ?? '';
    const moduleMatch = GO_MOD_MODULE_RE.exec(text);
    const notes = moduleMatch ? [`module ${moduleMatch[1]!}`] : [];
    const gosumPresent = pathExists(join(root, 'go.sum'));
    return {
      ecosystem: 'go',
      manager: 'go modules',
      lockfile: gosumPresent ? 'go.sum' : null,
      lockfile_present: gosumPresent,
      manifest_present: true,
      installed: Presence.UNKNOWN,
      dependency_count: null,
      dev_dependency_count: null,
      notes,
    };
  }

  override async inspectTests(root: string): Promise<string[]> {
    return pathExists(join(root, 'go.mod')) ? ['go test ./...'] : [];
  }

  override async inspectBuildCommands(root: string): Promise<string[]> {
    return pathExists(join(root, 'go.mod')) ? ['go build ./...'] : [];
  }

  override async healthChecks(_root: string, runtimes: RuntimeInfo[]): Promise<HealthIssue[]> {
    const issues: HealthIssue[] = [];
    for (const rt of runtimes) {
      if (rt.name === 'go' && rt.presence === Presence.NOT_INSTALLED) {
        issues.push({
          severity: Severity.CRITICAL,
          code: 'go.not_installed',
          title: 'Go runtime not found',
          message: 'No `go` executable was found on PATH.',
          evidence: ['which go -> not found'],
          recommendation: 'Install Go from https://go.dev/dl/ or your OS package manager.',
          confidence: null,
        });
      }
    }
    return issues;
  }
}
